File: tp3/TP03Q01/Lista.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class Lista {
  static Scanner scan = new Scanner(System.in);

  public class Personagem {
    static Scanner scan = new Scanner(System.in);
    static int comp = 0;
    private String id;
    private String name;
    private List<String> alternateNames;
    private String house;
    private String ancestry;
    private String species;
    private String patronus;
    private boolean hogwartsStaff;
    private boolean hogwartsStudent;
    private String actorName;
    private boolean alive;
    private LocalDate dateOfBirth;
    private int yearOfBirth;
    private String eyeColour;
    private String gender;
    private String hairColour;
    private boolean wizard;

    public Personagem() {
      this.id = "";
      this.name = "";
      this.alternateNames = new ArrayList<>();
      this.house = "";
      this.ancestry = "";
      this.species = "";
      this.patronus = "";
      this.hogwartsStaff = true;
      this.hogwartsStudent = true;
      this.actorName = "";
      this.alive = true;
      this.dateOfBirth = LocalDate.of(0000, 1, 1);
      this.yearOfBirth = -1;
      this.eyeColour = "";
      this.gender = "";
      this.hairColour = "";
      this.wizard = true;
    }

    public Personagem(String x) {
      buscarPorId(x, "/tmp/characters.csv");
    }

    public String getId() {
      return id;
    }

    public void setId(String id) {
      this.id = id;
    }

    public String getNome() {
      return name;
    }

    public void setNome(String nome) {
      this.name = nome;
    }

    public List<String> getAlt() {
      return alternateNames;
    }

    public void setAlt(List<String> alt) {
      this.alternateNames = alt;
    }

    public String getCasa() {
      return house;
    }

    public void setCasa(String casa) {
      this.house = casa;
    }

    public String getAncestry() {
      return ancestry;
    }

    public void setAncestry(String ancestralidade) {
      this.ancestry = ancestralidade;
    }

    public String getEspecie() {
      return species;
    }

    public void setEspecie(String especie) {
      this.species = especie;
    }

    public String getPatrono() {
      return patronus;
    }

    public void setPatrono(String patrono) {
      this.patronus = patrono;
    }

    public boolean getStaff() {
      return hogwartsStaff;
    }

    public void setStaff(boolean staff) {
      this.hogwartsStaff = staff;
    }

    public boolean getEstudante() {
      return hogwartsStudent;
    }

    public void setEstudante(boolean estudante) {
      this.hogwartsStudent = estudante;
    }

    public String getAtor() {
      return actorName;
    }

    public void setAtor(String ator) {
      this.actorName = ator;
    }

    public boolean getVivo() {
      return hogwartsStaff;
    }

    public void setVivo(boolean vivo) {
      this.alive = vivo;
    }

    public LocalDate getData() {
      return dateOfBirth;
    }

    public void setData(LocalDate data) {
      this.dateOfBirth = data;
    }

    public int getAno() {
      return yearOfBirth;
    }

    public void SetAno(int ano) {
      this.yearOfBirth = ano;
    }

    public String getOlho() {
      return eyeColour;
    }

    public void setOlho(String olho) {
      this.eyeColour = olho;
    }

    public String getGenero() {
      return gender;
    }

    public void setGenero(String genero) {
      this.gender = genero;
    }

    public String getCabelo() {
      return hairColour;
    }

    public void setCabelo(String cabelo) {
      this.hairColour = cabelo;
    }

    public boolean getBruxo() {
      return wizard;
    }

    public void setBruxo(boolean bruxo) {
      this.wizard = bruxo;
    }

    protected Personagem clone() throws CloneNotSupportedException {
      Personagem novo = new Personagem();
      novo.id = this.id;
      novo.name = this.name;
      novo.alternateNames = this.alternateNames;
      novo.house = this.house;
      novo.ancestry = novo.ancestry;
      novo.species = this.species;
      novo.patronus = this.patronus;
      novo.hogwartsStaff = this.hogwartsStaff;
      novo.hogwartsStudent = this.hogwartsStudent;
      novo.actorName = this.actorName;
      novo.alive = this.alive;
      novo.dateOfBirth = this.dateOfBirth;
      novo.yearOfBirth = this.yearOfBirth;
      novo.eyeColour = this.eyeColour;
      novo.gender = this.gender;
      novo.hairColour = this.hairColour;
      novo.wizard = this.wizard;
      return novo;
    }

    public void ler(String nomeArquivo) throws IOException {
      FileReader file = new FileReader(nomeArquivo);
      BufferedReader buffer = new BufferedReader(file);

      String line;
      boolean firstLine = true;

      while ((line = buffer.readLine()) != null) {
        if (firstLine) {
          firstLine = false;
          continue;
        }

        String[] fields = line.split(",");
        if (fields.length > 0) {
          this.id = fields.length > 0 ? fields[0] : "nao informado";
          this.name = fields.length > 1 ? fields[1] : "nao informado";
          this.alternateNames = new ArrayList<>();
          if (fields.length > 2 && !fields[2].isEmpty()) {
            String[] alternateNamesArray = fields[2].substring(1, fields[2].length() - 1).split(", ");
            this.alternateNames.addAll(Arrays.asList(alternateNamesArray));
          }
          this.house = fields.length > 3 ? fields[3] : "nao informado";
          this.ancestry = fields.length > 4 ? fields[4] : "nao informado";
          this.species = fields.length > 5 ? fields[5] : "nao informado";
          this.patronus = fields.length > 6 ? fields[6] : "nao informado";
          if (fields.length > 7 && fields[7].compareTo("VERDADEIRO") == 0) {
            this.hogwartsStaff = true;
          } else {
            this.hogwartsStaff = false;
          }
          if (fields.length > 8 && fields[8].compareTo("VERDADEIRO") == 0) {
            this.hogwartsStudent = true;
          } else {
            this.hogwartsStudent = false;
          }
          this.actorName = fields.length > 9 ? fields[9] : "nao informado";
          if (fields.length > 10 && fields[10].compareTo("VERDADEIRO") == 0) {
            this.alive = true;
          } else {
            this.alive = false;
          }
          DateTimeFormatter formatter1 = DateTimeFormatter.ofPattern("dd-MM-yyyy");
          DateTimeFormatter formatter2 = DateTimeFormatter.ofPattern("d-M-yyyy");
          try {
            this.dateOfBirth = fields.length > 12 ? LocalDate.parse(fields[12], formatter1) : null;
          } catch (DateTimeParseException e) {
            this.dateOfBirth = fields.length > 12 ? LocalDate.parse(fields[12], formatter2) : null;
          }
          this.yearOfBirth = fields.length > 13 ? Integer.parseInt(fields[13]) : -1;
          this.eyeColour = fields.length > 14 ? fields[14] : "nao informado";
          this.gender = fields.length > 15 ? fields[15] : "nao informado";
          this.hairColour = fields.length > 16 ? fields[16] : "nao informado";
          if (fields.length > 17 && fields[17].compareTo("VERDADEIRO") == 0) {
            this.wizard = true;
          } else {
            this.wizard = false;
          }
        }
      }

      buffer.close();
      file.close();
    }

    public void buscarPorId(String characterId, String nomeArquivo) {
      try {
        FileReader file = new FileReader(nomeArquivo);
        BufferedReader buffer = new BufferedReader(file);

        String line;
        boolean firstLine = true;

        while ((line = buffer.readLine()) != null) {
          if (firstLine) {
            firstLine = false;
            continue;
          }

          String[] fields = line.split(";");

          if (fields.length > 0 && fields[0].equals(String.valueOf(characterId))) {
            this.id = fields.length > 0 ? fields[0] : "nao informado";
            this.name = fields.length > 1 ? fields[1] : "nao informado";
            this.alternateNames = new ArrayList<>();
            if (fields.length > 2 && !fields[2].isEmpty()) {
              String[] alternateNamesArray = fields[2].substring(1, fields[2].length() - 1).split(", ");
              this.alternateNames.addAll(Arrays.asList(alternateNamesArray));
            }
            this.house = fields.length > 3 ? fields[3] : "nao informado";
            this.ancestry = fields.length > 4 ? fields[4] : "nao informado";
            this.species = fields.length > 5 ? fields[5] : "nao informado";
            this.patronus = fields.length > 6 ? fields[6] : "nao informado";
            if (fields.length > 7 && fields[7].compareTo("VERDADEIRO") == 0) {
              this.hogwartsStaff = true;
            } else {
              this.hogwartsStaff = false;
            }
            if (fields.length > 8 && fields[8].compareTo("VERDADEIRO") == 0) {
              this.hogwartsStudent = true;
            } else {
              this.hogwartsStudent = false;
            }
            this.actorName = fields.length > 9 ? fields[9] : "nao informado";
            if (fields.length > 10 && fields[10].compareTo("VERDADEIRO") == 0) {
              this.alive = true;
            } else {
              this.alive = false;
            }
            DateTimeFormatter formatter1 = DateTimeFormatter.ofPattern("dd-MM-yyyy");
            DateTimeFormatter formatter2 = DateTimeFormatter.ofPattern("d-M-yyyy");
            try {
              this.dateOfBirth = fields.length > 12 ? LocalDate.parse(fields[12], formatter1) : null;
            } catch (DateTimeParseException e) {
              this.dateOfBirth = fields.length > 12 ? LocalDate.parse(fields[12], formatter2) : null;
            }
            this.yearOfBirth = fields.length > 13 ? Integer.parseInt(fields[13]) : -1;
            this.eyeColour = fields.length > 14 ? fields[14] : "nao informado";
            this.gender = fields.length > 15 ? fields[15] : "nao informado";
            this.hairColour = fields.length > 16 ? fields[16] : "nao informado";
            if (fields.length > 17 && fields[17].compareTo("VERDADEIRO") == 0) {
              this.wizard = true;
            } else {
              this.wizard = false;
            }
            break;
          }
        }
        buffer.close();
        file.close();
      } catch (IOException e) {
        e.printStackTrace();
      }
    }

    public void imprimir() {
      System.out.println(toString());
    }

    public String toString() {
      DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
      String formattedDate = dateOfBirth.format(formatter);
      String alternateNamesString = alternateNames.toString().replace("[", "{").replace("]", "}").replace("'", "");
      String resp = "[" + id + " ## " + name + " ## " + alternateNamesString + " ## " + house + " ## ";
      resp += ancestry + " ## " + species + " ## ";
      resp += patronus + " ## " + hogwartsStaff + " ## ";
      resp += hogwartsStudent + " ## " + actorName + " ## ";
      resp += alive + " ## " + formattedDate + " ## ";
      resp += yearOfBirth + " ## " + eyeColour + " ## ";
      resp += gender + " ## " + hairColour + " ## ";
      resp += wizard + "]";
      return resp;
    }

    public void imprimirPosition(int i) {
      System.out.println(toStringWithPosition(i));
    }

    public String toStringWithPosition(int position) {
      DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
      String formattedDate = dateOfBirth.format(formatter);
      String alternateNamesString = alternateNames.toString().replace("[", "{").replace("]", "}").replace("'", "");
      String resp = "[" + position + " ## " + id + " ## " + name + " ## " + alternateNamesString + " ## " + house
          + " ## ";
      resp += ancestry + " ## " + species + " ## ";
      resp += patronus + " ## " + hogwartsStaff + " ## ";
      resp += hogwartsStudent + " ## " + actorName + " ## ";
      resp += alive + " ## " + formattedDate + " ## ";
      resp += yearOfBirth + " ## " + eyeColour + " ## ";
      resp += gender + " ## " + hairColour + " ## ";
      resp += wizard + "]";
      return resp;
    }

    public void imprimirNome() {
      System.out.println(name);
    }
  }

  int n, total;
  Personagem[] arr;

  public Lista() {
    n = 0;
    total = 10000;
    arr = new Personagem[total];
  }

  public void inserirInicio(Personagem x) {
    this.n++;
    for (int i = n - 1; i >= 1; i--) {
      this.arr[i] = this.arr[i - 1];
    }
    this.arr[0] = x;
  }

  public void inserirFim(Personagem x) {
    arr[n] = x;
    n++;
  }

  public void removerInicio() {
    System.out.println("(R) " + arr[0].getNome());
    for (int i = 0; i < n - 1; i++) {
      arr[i] = arr[i + 1];
    }
    n--;
  }

  public void removerFim() {
    if (n > 0) {
      System.out.println("(R) " + arr[n - 1].getNome());
      n--;
    }
  }

  public void inserir(Personagem x, int pos) {
    for (int i = n - 1; i >= pos; i--) {
      arr[i + 1] = arr[i];
    }
    arr[pos] = x;
    n++;
  }

  public void remover(int pos) {
    System.out.println("(R) " + arr[pos].getNome());
    for (int i = pos; i < n - 1; i++) {
      arr[i] = arr[i + 1];
    }

    n--;
  }

  public void mostrar() {
    for (int i = 0; i < n; i++) {
      arr[i].imprimirPosition(i);
      ;
    }
  }

  public void metodos() {
    String linha, metodo;
    linha = scan.nextLine();
    String[] words = linha.split(" ");
    metodo = words[0];
    // System.out.println("Metodo: " + metodo);
    if (metodo.equals("II")) {
      String id = words[1];
      // System.out.println("ID:" + id);
      inserirInicio(new Personagem(id));
    }
    if (metodo.equals("IF")) {
      String id = words[1];
      // System.out.println("ID: " + id);
      inserirFim(new Personagem(id));
    }
    if (metodo.equals("I*")) {
      int pos = Integer.valueOf(words[1]);
      String id = words[2];
      // System.out.println("Pos: " + pos + "ID:" + id);
      inserir(new Personagem(id), pos);
    }
    if (metodo.equals("RI")) {
      removerInicio();
    }
    if (metodo.equals("RF")) {
      removerFim();
    }
    if (metodo.equals("R*")) {
      int pos = Integer.valueOf(words[1]);
      // System.out.println("Pos: " + pos);
      remover(pos);
    }
  }

  public static void main(String[] args) {
    Lista lista = new Lista();
    String linha;
    linha = scan.nextLine();
    // System.out.println("Linha: " + linha);
    while (!linha.equals("FIM")) {
      lista.inserirFim(lista.new Personagem(linha));
      // lista.mostrar();
      linha = scan.nextLine();
      // System.out.println("Linha: " + linha);
    }
    int n = scan.nextInt();
    scan.nextLine();
    // System.out.println("N: " + n);
    for (int i = 0; i < n; i++) {
      lista.metodos();
    }
    lista.mostrar();
  }
}

File: tp3/TP03Q02/lista.c
#include <stdbool.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <ctype.h>

typedef struct
{
    char id[100];
    char name[100];
    char alternateNames[5][100];
    char house[100];
    char ancestry[100];
    char species[100];
    char patronus[100];
    bool hogwartsStaff;
    bool hogwartsStudent;
    char actorName[100];
    bool alive;
    char dateOfBirth[100];
    char yearOfBirth[100];
    char eyeColour[100];
    char gender[100];
    char hairColour[100];
    int wizard;
} Personagem;

void replace(char *str, char *find, char *replace)
{
    int find_len = strlen(find);
    int replace_len = strlen(replace);
    char *tmp = strstr(str, find);
    while (tmp != NULL)
    {
        memmove(tmp + replace_len, tmp + find_len, strlen(tmp + find_len) + 1);
        memcpy(tmp, replace, replace_len);
        tmp = strstr(tmp + replace_len, find);
    }
}

int cmpStrings(char *str1, char *str2)
{
    int i = 0;
    while (str1[i] != '\0' && str2[i] != '\0')
    {
        if (str1[i] > str2[i])
        {
            return 1;
        }
        else if (str1[i] < str2[i])
        {
            return -1;
        }
        i++;
    }
    if (str1[i] == '\0' && str2[i] != '\0')
    {
        return -1;
    }
    else if (str1[i] != '\0' && str2[i] == '\0')
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

void limparString(char str[])
{
    int i, j = 0;
    for (i = 0; str[i] != '\0'; i++)
    {
        if (!isspace(str[i]))
        {
            str[j++] = str[i];
        }
    }
    str[j] = '\0';
}

void ler(Personagem personagens[], char *arquivo)
{
    FILE *file = fopen(arquivo, "r");

    char linha[1050];
    int i = 0;
    while (fgets(linha, sizeof(linha), file))
    {

        replace(linha, ";;", ";*;");
        char *field = strtok(linha, ";");
        strcpy(personagens[i].id, field);
        field = strtok(NULL, ";");

        strcpy(personagens[i].name, field);
        field = strtok(NULL, ";");

        field++;
        char *endfield = strchr(field, ']');
        if (endfield != NULL)
        {
            *endfield = '\0';
            strcpy(personagens[i].alternateNames, field);
        }
        else
        {
            strcpy(personagens[i].alternateNames, "");
        }
        field = strtok(NULL, ";");

        strcpy(personagens[i].house, field);
        field = strtok(NULL, ";");

        strcpy(personagens[i].ancestry, field);
        field = strtok(NULL, ";");

        strcpy(personagens[i].species, field);
        field = strtok(NULL, ";");

        strcpy(personagens[i].patronus, field);
        field = strtok(NULL, ";");

        personagens[i].hogwartsStaff = strcmp(field, "VERDADEIRO") == 0 ? true : false;
        field = strtok(NULL, ";");

        personagens[i].hogwartsStudent = strcmp(field, "VERDADEIRO") == 0 ? true : false;
        field = strtok(NULL, ";");

        strcpy(personagens[i].actorName, field);
        field = strtok(NULL, ";");

        personagens[i].alive = strcmp(field, "VERDADEIRO") == 0 ? true : false;
        field = strtok(NULL, ";");
        field = strtok(NULL, ";");

        if (strcmp("23-6-1980", field) == 0)
        {
            strcpy(personagens[i].dateOfBirth, "23-06-1980");
        }
        else
            strcpy(personagens[i].dateOfBirth, field);
        field = strtok(NULL, ";");

        strcpy(personagens[i].yearOfBirth, field);
        field = strtok(NULL, ";");

        strcpy(personagens[i].eyeColour, field);
        field = strtok(NULL, ";");

        strcpy(personagens[i].gender, field);
        field = strtok(NULL, ";");

        strcpy(personagens[i].hairColour, field);
        field = strtok(NULL, ";");

        limparString(field);
        if (strcmp(field, "VERDADEIRO") == 0)
        {
            personagens[i].wizard = 1;
        }
        else
        {
            personagens[i].wizard = 0;
        }
        i++;
    }
    fclose(file);
}
void imprimirMensagemRemocao(char *nome) { printf("(R) %s\n", nome); }

#define MAX_SIZE 500

typedef struct {
  Personagem arr[MAX_SIZE];
  int n;
  int total;
} Lista;

Lista newLista() {
  Lista lista;
  lista.n = 0;
  lista.total = 500;
  return lista;
}

Lista inserirInicio(Lista lista, Personagem personagem) {
  for (int i = lista.n; i > 0; i--) {
    lista.arr[i] = lista.arr[i - 1];
  }
  lista.arr[0] = personagem;
  lista.n++;
  return lista;
}

Lista inserirFim(Lista lista, Personagem personagem) {
  lista.arr[lista.n] = personagem;
  lista.n++;
  return lista;
}

Lista removerInicio(Lista lista) {
  if (lista.n > 0) {
    imprimirMensagemRemocao(lista.arr[0].name);
    for (int i = 0; i < lista.n - 1; i++) {
      lista.arr[i] = lista.arr[i + 1];
    }
    lista.n--;
  }
  return lista;
}

Lista removerFim(Lista lista) {
  if (lista.n > 0) {
    imprimirMensagemRemocao(lista.arr[lista.n - 1].name);
    lista.n--;
  }
  return lista;
}

Lista inserir(Lista lista, Personagem Personagem, int pos) {
  for (int i = lista.n; i > pos; i--) {
    lista.arr[i] = lista.arr[i - 1];
  }
  lista.arr[pos] = Personagem;
  lista.n++;
  return lista;
}

Lista remover(Lista lista, int pos) {
  if (pos >= 0 && pos < lista.n) {
    imprimirMensagemRemocao(lista.arr[pos].name);
    for (int i = pos; i < lista.n - 1; i++) {
      lista.arr[i] = lista.arr[i + 1];
    }
    lista.n--;
  }
  return lista;
}

Personagem findPersonagemById(Personagem personagens [], int n, char id []) {
  for (int i = 0; i < n; i++) {
    if (strcmp(personagens[i].id, id) == 0) {
      return personagens[i];
    }
  }
}

void metodos(Lista lista, Personagem personagens [], int n) {
  char linha[30];
  fgets(linha, 30, stdin);
  char *metodo = strtok(linha, " ");
  if (strcmp(metodo, "II") == 0) {
    char *id = strtok(NULL, " ");
    Personagem p = findPersonagemById(personagens, n, id);
    inserirInicio(lista, p);
  } else if (strcmp(metodo, "IF") == 0) {
    char *id = strtok(NULL, " ");
    Personagem p = findPersonagemById(personagens, n, id);
      inserirFim(lista, p);
  } else if (strcmp(metodo, "I*") == 0) {
    char *pos_str = strtok(NULL, " ");
    int pos = atoi(pos_str);
    char *id = strtok(NULL, " ");
    Personagem p = findPersonagemById(personagens, n, id);
    inserir(lista, p, pos);
  } else if (strcmp(metodo, "RI") == 0) {
    removerInicio(lista);
  } else if (strcmp(metodo, "RF") == 0) {
    removerFim(lista);
  } else if (strcmp(metodo, "R*") == 0) {
    char *pos_str = strtok(NULL, " ");
    int pos = atoi(pos_str);
    remover(lista, pos);
  }
}

void mostrar(Lista lista) {
  for (int i = 0; i < lista.n; i++) {
    printf("[%d ", i);
    printf("## %s ## ", lista.arr[i].name);
    printf("## %s ## ", lista.arr[i].house);
    printf("## %s ## ", lista.arr[i].patronus);
    printf("## %s ## ", lista.arr[i].species);
    printf("## %s ## ", lista.arr[i].dateOfBirth);
    printf("## %s ## ", lista.arr[i].yearOfBirth);
    printf("## %s ## ", lista.arr[i].eyeColour);
    printf("## %s ## ", lista.arr[i].gender);
    printf("## %s ## ", lista.arr[i].hairColour);
    printf("## %d ]\n", lista.arr[i].wizard);
    }
}



int main(void) {
  Personagem personagens [500];
  Lista lista = newLista();
  char linha [50];
  ler(personagens, "characters.csv");
  scanf("%s", linha);
  printf("Linha: %s\n", linha);
  while (strcmp(linha, "FIM") != 0) {
    Personagem p = findPersonagemById(personagens, 500, linha);
    printf("Personagem: %s\n", p.name);
    inserirFim(lista, p);
    mostrar(lista);
    scanf("%s", linha);
    printf("Linha: %s\n", linha);
  }
  int n;
  scanf("%d", &n);
  printf("N: %d\n", n);
  for (int i = 0; i < n; i++) {
    metodos(lista, personagens,n);
  } 
  mostrar(lista);
  return 0;
}